### DISCUSSION TITLE
Access Dipole array

### DISCUSSION THREAD

**Discussion by wagdy88** [jonescompneurolab/hnn-core#505] [Q&A] [answered]
Hello, 
Is there a way to access the Dipole's array? I'd like to access - and operate on- the arrays for aggregates, L2, and L5. 

Thanks 
Mohamed

**Comment by jasmainak** [comment, marked as answer]
Did you look at the [documentation](https://jonescompneurolab.github.io/hnn-core/stable/generated/hnn_core.dipole.Dipole.html#hnn_core.dipole.Dipole) ? ;-)

You can do:

```py
dipole.data['L2']
dipole.data['L5']
dipole.data['agg']
```

**Reply by wagdy88** [reply]
Hi @jasmainak , thanks for sending this over. A follow up question, I am trying to plot the new dipole I created after making changes through dipole.data. However, I get a message 'numpy.float64' object has no attribute 'scale_applied'. Is there a way to fix this?
Thanks 
Mohamed

**Reply by wagdy88** [reply]
Here's what I am trying to do, I am trying to subtract the mean from the dipole value and analyze the outcome using power spectrum density.

**Reply by jasmainak** [reply]
A little hard to debug with the information you provided to me. The `Dipole` object has an attribute called `scale_applied` which should be `1.` by default by can be larger if you assume that the HNN network represents a larger network of cells than 270 :) Can you share the commands you used?

**Reply by wagdy88** [reply]
Thanks @jasmainak. I want to keep the network as is i.e. 270 cells. Here is what I have been trying to do:
We are trying to quantify the gamma signals using PSD. When we plotted the original dipole, it seemed that the signals are shifted up around the zero. So, our thought was to correct that by subtracting the mean dipole from the original dipole and then re-plot it.

Here are the commands I used:

dpl_dataL2 = dpl.data['L2']
dpl_dataL2_avg = dpl.data['L2'].mean()
dpl_dataL2_modified = dpl_dataL2 - dpl_dataL2_avg

plot_dipole(dpl_dataL2_modified)
and that's when I get the message " 'numpy.float64' object has no attribute 'scale_applied'"

**Reply by jasmainak** [reply]
this is easy to fix ;-) You passed an array instead of Dipole object to `plot_dipole`. This merits a check in our function. You should do:

```py
dpl_new = Dipole(dpl.times, dpl_dataL2_modified)
dpl_new.plot()
```

**Reply by wagdy88** [reply]
Thanks! I'll try this :-)

On the other hand, if the issue seems doable for a new contributor, I'd be happy to work on it

**Reply by jasmainak** [reply]
I think it's totally doable. You will mostly have to look around in the code how we do type checks and copy that.

**Comment by jasmainak** [comment]
There is also a private function [_baseline_renormalize](https://github.com/jonescompneurolab/hnn-core/blob/b974fed4317b9b03ec732850829d9c17841c6ee2/hnn_core/dipole.py#L517) in the code but it seems that at some point we stopped using it ... @rythorpe @ntolley @cjayb any clue, opinion?

**Reply by rythorpe** [reply]
I think we stopped using it in #264 because it really only made sense for a very specific use-case (ERPs), and even then, it didn't do a good job. Note that `_baseline_renormalize` is still used in the tests that compare simulated waveforms between here and legacy HNN.

**Reply by rythorpe** [reply]
Correction: we still call `_baseline_renormalize` by default in [`_gather_trial_data()`](https://github.com/jonescompneurolab/hnn-core/blob/b1a81946a68297a6f315bf25bdd4915c922fa0ee/hnn_core/parallel_backends.py#L69).
